### PR TITLE
Added OpenCL.pc file.

### PR DESCRIPTION
**Pull request by Kerilk**
This adds pkg-config support to the OpenCL ICD Loader package.
This fixes https://github.com/KhronosGroup/OpenCL-ICD-Loader/issues/187

### CODE REVIEW

File: CMakeLists.txt
Comment (bashbaug) [inline CMakeLists.txt:215]: Confirming: Do we only want to do this for UNIX?  Or should we do this for other operating systems also?

File: OpenCL.pc.in
Comment (bashbaug) [inline OpenCL.pc.in:6]: Confirming: It's OK if this version number is different than our SOVERSION?

If so, this is another place that would be great to automate with #182!

**Review by bashbaug — COMMENTED** [review]
I'm not a pkgconfig expert so please bear with my (probably basic) questions.

Should we also have a dependency on a similar OpenCL headers package?

File: OpenCL.pc.in
Comment (Kerilk) [reply OpenCL.pc.in:6]: > Confirming: It's OK if this version number is different than our SOVERSION?

This mimics what `ocl-icd` does here: https://github.com/OCL-dev/ocl-icd/blob/master/OpenCL.pc.in where @OPENCL_VERSION@ is currently evaluated to `3.0`

> If so, this is another place that would be great to automate with https://github.com/KhronosGroup/OpenCL-ICD-Loader/pull/182!

Indeed

File: CMakeLists.txt
Comment (Kerilk) [reply CMakeLists.txt:215]: I was unsure if these could successfully be installed for a regular Windows build so erred on the side of caution here. 
I found this to confirm my bias:
https://github.com/pocl/pocl/blob/63b28753749d16209db54614e6a957bd5a203087/CMakeLists.txt#L1720-L1726
but we may reevaluate. This `UNIX` variable should cover Linux, MacOS, cygwin. I am unsure about MinGW and Android.

Comment (jenatali) [reply CMakeLists.txt:215]: PkgConfig files can be used on Windows (though it's somewhat uncommon). https://sourceforge.net/projects/pkgconfiglite/ can be used on Windows to search and use them.

**Comment by Kerilk** [comment]
> I'm not a pkgconfig expert so please bear with my (probably basic) questions.

My knowledge is very basic as well.

> Should we also have a dependency on a similar OpenCL headers package?

Excellent  question. Here, the .pc file is for both library and headers assuming both are installed using the same prefix. This is the same in `ocl-icd`. I agree that conceptually, the headers and the loader should be kept separate. In practice, installing the two alongside each other makes sense, as building and linking requires both.
In debian the the .pc file is packaged in the `ocl-icd-opencl-dev` package:
https://packages.debian.org/bullseye/amd64/ocl-icd-opencl-dev/filelist
Which contains the .so symlink and the .pc file.

I assume we could split it in two, one for the headers and one for the library and make the latter one depend on the former. I'll try experimenting with this.

File: OpenCL.pc.in
Comment (Kerilk) [inline OpenCL.pc.in:8]:
```suggestion
libdir=@CMAKE_INSTALL_FULL_LIBDIR@

Name: OpenCL
Description: Khronos OpenCL ICD Loader
Requires: OpenCL-Headers
Version: 3.0
Libs: -L${libdir} -lOpenCL
```
This would work in conjunction with: https://github.com/KhronosGroup/OpenCL-Headers/pull/213 that would need to be merged before adopting this change.

**Comment by bashbaug** [comment]
Probably a good idea to get input from @MathiasMagnus before we go too far.  I'm not sure if it makes sense to have multiple pkgconfig files for each of the OpenCL components, or a single pkgconfig file for the OpenCL SDK, or both.

File: CMakeLists.txt
Comment (Kerilk) [reply CMakeLists.txt:215]: @jenatali thanks for confirming. Is there a convention we need to follow as far as the install location is concerned?

Comment (jenatali) [reply CMakeLists.txt:215]: No, generally whoever is building a PkgConfig enabled project will specify the install prefix on the command line. From there, normal UNIX subdirectories are expected.

File: CMakeLists.txt
Comment (Kerilk) [inline CMakeLists.txt:221]:
```suggestion
configure_file(OpenCL.pc.in OpenCL.pc @ONLY)
set(pkg_config_location ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
install(
  FILES ${CMAKE_CURRENT_BINARY_DIR}/OpenCL.pc
  DESTINATION ${pkg_config_location})
```
This removes the Unix check to enable `pkg-config` support for Windows.

**Comment by Kerilk** [comment]
> Probably a good idea to get input from @MathiasMagnus before we go too far. I'm not sure if it makes sense to have multiple pkgconfig files for each of the OpenCL components, or a single pkgconfig file for the OpenCL SDK, or both.

Indeed. I tested and split the .pc file between the Headers and Loader, and it worked as expected. I made a PR for the Header as well should we want to take this route: https://github.com/KhronosGroup/OpenCL-Headers/pull/213

File: CMakeLists.txt
Comment (mmuetzel) [reply CMakeLists.txt:221]: Many projects are using `if (NOT MSVC)` as a guard for .pc file creation.

File: OpenCL.pc.in
Comment (mmuetzel) [inline OpenCL.pc.in:2]: Proposed change to allow easier relocation:
``` diff
-libdir=@CMAKE_INSTALL_FULL_LIBDIR@
-includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@
+prefix=@CMAKE_INSTALL_PREFIX@
+exec_prefix=${prefix}
+libdir=${exec_prefix}/lib
+includedir=${prefix}/include
```

Comment (Kerilk) [reply OpenCL.pc.in:2]: Thanks for the suggestion. If we decide to keep a single .pc file, rather than split the headers, I will most probably merge this. If we split the .pc files, then only `libdir` is used, and in this case I'd rather use the full version.

File: CMakeLists.txt
Comment (Kerilk) [reply CMakeLists.txt:221]: Thanks for the suggestion! Is there a reason not to install it on windows (see above discussion)? We could also make it optional and off by default on Windows, we would need to document that.

File: OpenCL.pc.in
Comment (mmuetzel) [reply OpenCL.pc.in:2]: Ok. The most important change wrt to relocation would be to keep `prefix` separate. Whether or not to include `includedir` doesn't matter for that question.

Comment (Kerilk) [reply OpenCL.pc.in:2]: Do you want to do a formal suggestion so that I can merge it? I will merge it right away.

File: CMakeLists.txt
Comment (Kerilk) [inline CMakeLists.txt:215]:
```suggestion
if (NOT MSVC)
```

Comment (Kerilk) [reply CMakeLists.txt]: For now, according to @mmuetzel suggestion, I deactivated pkg-config support when MSVC is detected.

File: OpenCL.pc.in
Comment (mmuetzel) [reply OpenCL.pc.in:2]: I don't know how to do that. Is that a PR to your branch?
If so: https://github.com/Kerilk/OpenCL-ICD-Loader/pull/1

Comment (mmuetzel) [inline OpenCL.pc.in:2]:
``` suggestion
prefix=@CMAKE_INSTALL_PREFIX@
exec_prefix=${prefix}
libdir=${exec_prefix}/@CMAKE_INSTALL_LIBDIR@
includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@
```

Think I figured it out

Comment (Kerilk) [reply OpenCL.pc.in:2]: Thanks!

File: OpenCL.pc.in
Comment (Kerilk) [reply OpenCL.pc.in]: Outdated

File: OpenCL.pc.in
Comment (Kerilk) [reply OpenCL.pc.in]: Thanks for the contribution, merged below.

File: OpenCL.pc.in
Comment (Kerilk) [inline OpenCL.pc.in:7]:
```suggestion

Name: OpenCL
Description: Khronos OpenCL ICD Loader
Requires: OpenCL-Headers
```
This would work in conjunction with: https://github.com/KhronosGroup/OpenCL-Headers/pull/213 that would need to be merged before adopting this change.

**Comment by MathiasMagnus** [comment]
I have less experience with pkg-config (because I mostly use Windows), but I do understand some users would appreciate this.

My only input here is that whatever type of consumption we envision using `.pc` files, we should test in CI. There are [consumption tests](https://github.com/KhronosGroup/OpenCL-ICD-Loader/blob/b7a648b2702e5484725163dcb99a3960370807b5/.github/workflows/linux.yml#L232-L248) in CI where we test that we don't forget to install anything that users will be looking for inside build trees, and that the contents of the install tree are usable. If we start deploying `.pc` files, I think we should test that their contents are usable and they are deployed in the proper places (and that it stays that way as the repo progresses).

I would feel comfortable approving once CI is in place for this. We can do so without it and say it's a "best effort" basis, but that may invite lots of issues from people wanting to use it in territories which we could otherwise cover in CI.

**Comment by mmuetzel** [comment]
A test in CI could maybe look similar to something like this:
```
PKG_CONFIG_PATH="/path/to/where/OpenCL.pc/was/installed" pkg-config --libs OpenCL
PKG_CONFIG_PATH="/path/to/where/OpenCL.pc/was/installed" pkg-config --cflags OpenCL
```
It might also be worth to check what those commands return. But that might depend a lot on the platform, install prefix, ...

File: .github/workflows/linux.yml
Comment (mmuetzel) [inline .github/workflows/linux.yml]: Same typo like originally in the macOS job definition: `--clfags` -> `--cflags`

File: .github/workflows/linux.yml
Comment (mmuetzel) [inline .github/workflows/linux.yml]: See above.

Comment (mmuetzel) [reply .github/workflows/linux.yml]: Oops. Didn't notice that this is already fixed. Please, ignore.

**Review by MathiasMagnus — APPROVED** [review]
I personally would've aimed at tests which don't just run a regular expression on the output of pkg-config, but actually try to compile something with them. The same `pkgconfig.c` source file can be used for that which we have both here and in the OpenCL-Headers repo. pkg-config also exists on Windows, albeit nobody uses it, so the lack of support for that shouldn't be a dealbreaker. We may want to open an issue so we don't lose sight of that feature missing.

Otherwise I wouldn't want to block this, as it's already useful to some, and we do have deployment tests for the `.pc` files so at least that won't be broken by mistake, and the c

**Comment by jenatali** [comment]
> pkg-config also exists on Windows, albeit nobody uses it

I do just want to point out that Meson, which is typically used for Linux projects, does also have pretty decent Windows support, and does use pkg-config as one of its core dependency lookup methods. So, yeah we do actually use pkg-config on Windows for a few dependencies, specifically libclc from LLVM, and SPIRV-Tools. That said, I don't have a stake in whether this project supports it on Windows or not.

**Comment by Kerilk** [comment]
@MathiasMagnus, @jenatali thanks for the feedback. As I see it we have 3 remaining questions:
 1) Should we support it on Windows? I am all in favor but need help doing it right and testing it.
 2) Do we want the two .pc file or the single .pc file approach? I reached out to Andreas Beckmann who is very active in Debian OpenCL packaging for feedback on this issue. I am worried about causing issues for distribution maintainers.
 3) Do we want to add a build/run test to validate the content of the pkg-config file. I will try.

**Comment by jenatali** [comment]
Sure, I can help test it.

**Comment by Kerilk** [comment]
@MathiasMagnus any idea of what could be wrong with the cmake minimum? I don't see anything in the documentation that implies a change in the variables I am using.

**Comment by MathiasMagnus** [comment]
@Kerilk Take a look at [this PR](https://github.com/Kerilk/OpenCL-ICD-Loader/pull/2) to your branch. I believe it should solve the issue.

File: OpenCL.pc.in
Comment (eli-schwartz) [inline OpenCL.pc.in:4]: Previously, this used `CMAKE_INSTALL_FULL_*`, now it joins together `CMAKE_INSTALL_PREFIX` and `CMAKE_INSTALL_*`.

Actually, though, *both* approaches are kinda broken. Meson has a robust pkg-config generator that handles these edge cases and a lot more for you, but unfortunately CMake does not... however, you can work around this with a bit of scripting.

Please see https://github.com/jtojnar/cmake-snips/#assuming-cmake_install_dir-is-relative-path and including the very permissively licensed `JoinPaths.cmake` module, and using it before `configure_file()`.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: OpenCL.pc.in
Comment (Kerilk) [reply OpenCL.pc.in:4]: Thanks for the input. I was worrying about something like that, and it was the reason I tried going with `CMAKE_INSTALL_FULL_*` first. I will defer to our CMake expert (:wink: @MathiasMagnus) for how to integrate the above.

For reference, here is an example of a 'correct' integration:
https://github.com/fmtlib/fmt/pull/1657/files

Comment (mmuetzel) [reply OpenCL.pc.in:4]: Imho, the existing changes are "good enough" for the vast majority of use cases (that install headers and libraries at the same prefix).
Would it be ok to use the existing changes for now and fix the remaining use case (where headers or libraries are installed outside the prefix location) at a later time?

Comment (eli-schwartz) [reply OpenCL.pc.in:4]: On Windows you actually want to pretend that prefix is `${pcfiledir}\..\..` for as many times `\..` as there are relative links, but that's rapidly approaching and likely beyond the feasibility of what you can do without python or C++ ergonomics. If you want proper pkg-config support either use Meson or wait 20 years for cmake to decide they are willing to implement pkg-config generation? :shrug:

But actually I think windows should work fine with that module too, it allows `/` paths and cmake itself handles IS_ABSOLUTE (correctly, I expect).

So, the cmake snip isn't any worse on Windows than not using it.

...

Is this PR really being held up by my comment? I really, *really* doubt it.

Comment (eli-schwartz) [reply OpenCL.pc.in:4]: @Kerilk

> I will defer to our CMake expert (😉 @MathiasMagnus) for how to integrate the above.
> 
> For reference, here is an example of a 'correct' integration:
> https://github.com/fmtlib/fmt/pull/1657/files

Since you already know and give examples of integrating it, I don't see why you need a "cmake expert" to verify that it's okay to do.

Is cmake such a complicated build system that it's too dangerous to change anything whatsoever unless someone with a degree in cmake agrees with not only the what, but also the where and the how and the why?

I think it's safe enough to do it the obvious way :) and really, the best way to get a review from a cmake expert is to do an initial implementation and then ask for a ***review***. :)

Seems like the PR is currently just waiting on a known and expected change to be trivially implemented and pushed, so it would be great to get this moving forward again in the simple and obvious way. :)

**Comment by StuartDBrady** [comment]
I haven't followed the detail of the discussion here, but see `gl.pc` on a typical Linux distribution for an example of how this is done for GL, which typically has a dispatch layer similar to the ICD for OpenCL. Note that there is generally a consistent libdir and includedir under a single prefix, specified under a common pkgconfig file. I would strongly recommend talking to package maintainers before doing otherwise for OpenCL.

**Comment by Kerilk** [comment]
@StuartDBrady Thanks for the insights. I have reached out to the Debian OpenCL maintainer mailing list: `pkg-opencl-devel` . I someone with ties to other Linux distributions want to reach out as well, please do so.

> If this isn't sufficient for OpenCL, then I would ask why that is:
> 
>     * why isn't this a problem for GL and Vulkan too?
> 
>     * if it is a problem for GL and Vulkan, does this point at a more general problem with pkgconfig?
> 
>     * if it is a general problem with pkgconfig's design, then how is this worked around for other libraries?  Why should the OpenCL ICD and headers be the one special case where we don't expect a common prefix?  As stated, GL has a similar architecture.

I think in this case the difference comes from two aspects:
 - There exists 2 OpenCL ICD Loaders which, as they share library names cannot be installed in the same prefix. On distributions, you can solve the issue with `update-alternatives` if you wish them to coexist, or just rely on the package manager to switch them for you. If using environment modules this becomes more complicated.
 - OpenCL is used in the HPC world (more so than OpenGL and Vulkan for now) and it relies heavily on environment modules.

Other libraries have usually a single source for headers and libraries, or ship their own headers with their implementation (see MPI). If we wanted to take this route, then the headers should become a sub-module of the loader and always be installed side by side. This is not the choice we have currently made, and even not how distributions handle OpenCL headers and loaders as headers are shared between loaders (see https://packages.debian.org/stretch/opencl-dev for how it used to be done in debian, only ocl-icd remains nowadays)

File: OpenCL.pc.in
Comment (Kerilk) [inline OpenCL.pc.in:7]:
```suggestion

Name: OpenCL
Description: Khronos OpenCL ICD Loader
Requires: OpenCL-Headers
```
This would work in conjunction with: https://github.com/KhronosGroup/OpenCL-Headers/pull/213 that would need to be merged before adopting this change.

File: OpenCL.pc.in
Comment (Kerilk) [inline OpenCL.pc.in:10]:
```suggestion
```

**Comment by bashbaug** [comment]
Merging as discussed in the OpenCL teleconferences. This will give time for testing before the next release.

**Pull request merged**
